File: library/src/main/java/com/mohamedfadel91/mainthreadschedulerlib/MainThreadScheduler.java
package com.mohamedfadel91.mainthreadschedulerlib;

import android.os.Handler;
import android.os.Looper;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.test.espresso.Espresso;

/**
 * Created by fadel on 17/03/18.
 */
public class MainThreadScheduler {
    @Nullable
    static Handler mainThreadHandler;

    public static void scheduleWhenIdle(@NonNull final Runnable runnable){
        Runnable _runnable = new Runnable() {
            @Override
            public void run() {
                new Thread(new Runnable() {
                    @Override
                    public void run() {
                        Espresso.onIdle();
                        if(mainThreadHandler!=null) {
                            mainThreadHandler.post(runnable);
                        }
                    }
                }).start();
            }
        };

        /*
            check if current thread from mainThread just run the block
            otherwise it post it to the mainThread
         */
        if (Thread.currentThread() ==
                Looper.getMainLooper().getThread()) {
            _runnable.run();
        } else {
            if(mainThreadHandler!=null){
                mainThreadHandler.post(_runnable);
            }
        }
    }
}

File: library/src/main/java/com/mohamedfadel91/mainthreadschedulerlib/MainThreadScheduler.kt
package com.mohamedfadel91.mainthreadschedulerlib

import android.os.Handler
import android.os.Looper
import android.support.test.espresso.Espresso

/**
 * Created by fadel on 17/03/18.
 * This Class responsible for schedule functions or runnable
 * objects to run as soon as MainThread become idle..
 */
object MainThreadScheduler {
    internal var mainThreadHandler: Handler? = null

    /**
     * it post [unit] to run as soon as MainThread become Idle..
     */
    fun scheduleWhenIdle(unit: () -> Unit) {
        val block: () -> Unit = {
            Thread({
                Espresso.onIdle()
                mainThreadHandler?.post(unit)
            }).start()
        }

        /*
            check if current thread from mainThread just run the block
            otherwise it post it to the mainThread
         */
        if (Thread.currentThread() === Looper.getMainLooper().thread) {
            kotlin.run(block)
        } else {
            mainThreadHandler?.post(block)
        }
    }

    /**
     * it post [runnable] to run as soon as MainThread become Idle..
     */
    fun scheduleWhenIdle(runnable: Runnable) {
        val block: () -> Unit = {
            Thread({
                Espresso.onIdle()
                mainThreadHandler?.post(runnable)
            }).start()
        }

        /*
            check if current thread from mainThread just run the block
            otherwise it post it to the mainThread
         */
        if (Thread.currentThread() === Looper.getMainLooper().thread) {
            kotlin.run(block)
        } else {
            mainThreadHandler?.post(block)
        }
    }
}
